File: isbn-verifier/IsbnVerifier.cs
using System;
using System.Linq;
using System.Text.RegularExpressions;

public static class IsbnVerifier
{
    public static bool IsValid(string number)
    {
        var isbn = number.Replace("-", string.Empty);
        
        return HasValidFormat(isbn)
            ? IsValidIsbn10(isbn)
            : false;
    }

    private static bool HasValidFormat(string isbn) =>
        Regex.IsMatch(isbn, @"^(\d{9}[\dX])$");

    private static bool IsValidIsbn10(string isbn) =>
        isbn.Select(ToInt)
            .Select(ToIsbn10)
            .Sum() % 11 == 0;

    private static Func<char, int> ToInt => @char =>
    {
        if (@char == 'X')
            return 10;

        return Convert.ToInt32(@char.ToString());
    };

    private static Func<int, int, int> ToIsbn10 =>
        (digit, index) => digit * (10 - index);
}

File: fsharp/isbn-verifier/IsbnVerifier.fs
module IsbnVerifier

open System.Text.RegularExpressions

let [<Literal>] IsbnPattern = "^[0-9]{9}[0-9X]$"

let private digitToInt digit =
    match digit with
    | d when d = 'X' -> 10
    | _ -> int digit - int '0'

let multiply character digit =
    (10 - character) * (digit |> digitToInt) 

let private checkSum isbn = 
    isbn
    |> Seq.mapi multiply
    |> Seq.sum

let private cleanup (isbn: string) =
    isbn.Replace("-", "")

let isValid (isbn: string) = 
    let cleanedUpIsbn = isbn |> cleanup 

    match Regex.IsMatch(cleanedUpIsbn, IsbnPattern) with
    | false -> false
    | true  -> (cleanedUpIsbn |> checkSum) % 11 = 0